### ISSUE TITLE
@singleton() decorator no longer works as expected

### ISSUE THREAD

**Issue by AshleyGrant** [aurelia/framework#687]
**I'm submitting a bug report**

* **Library Version:**
1.0.8

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
6.9.1

* **NPM Version:**
3.10.8

* **JSPM OR Webpack AND Version**
Aurelia CLI 0.24.0

* **Browser:**
all 

* **Language:**
ESNext (haven't tested TypeScript)

**Current behavior:**
`@singleton` decorator does not causes pages loaded by the router to have a singleton VM. This is a regression that has happened since the last time @jdanyow's gist.run bundles were created, as this decorator works as expected there.
https://gist.run/?id=5826be4b6c4248f2aad40cf583a47c5c <-- this works as expected, but the same code run in an app created w/`0.24.0` of the CLI breaks

**Expected/desired behavior:**
VMs with the `@singleton()` decorator should only be created once.

**Comment by StrahilKazlachev** [comment]
It is not the decorator, probably also not the container.
- when `TemplatingRouteLoader.loadRoute` is called it [creates new child container](https://github.com/aurelia/templating-router/blob/1.0.1/src/route-loader.js#L15).
- then it calls [CompostionEngine.ensureViewModel](https://github.com/aurelia/templating-router/blob/1.0.1/src/route-loader.js#L33) using that child container for the `CompositionContext`.
- `.ensureViewModel` first calls [.autoRegister()](https://github.com/aurelia/templating/blob/1.2.0/src/composition-engine.js#L160) for the VM, on the child container.
- then it calls [.get()](https://github.com/aurelia/templating/blob/1.2.0/src/composition-engine.js#L166) on that child container for the VM. The container already has a registration for it so it just resolves it.

The VM is never registered in the root container. I'm not familiar with the `templating-router` but it seems like every time `TemplatingRouteLoader.loadRoute` is called a new child container is created, and everything repeats. @EisenbergEffect What do you think?

**Comment by AshleyGrant** [comment]
My man!

**Comment by AshleyGrant** [comment]
The thing is that this worked previously and has stopped working recently. I still need to look at the code and see what change caused the regression.

**Comment by StrahilKazlachev** [comment]
I used your sample and changed the `aurelia-dependency-injection` to 1.1.0 and now `One` gets instantiated once, and `Two` on every visit. I may have a wrong understanding but I did expect the same behavior for `Two` - since not having an explicit registration should make it a singleton in the root.

**Comment by EisenbergEffect** [comment]
It sounds like it's working correctly with the latest then. Can this be closed?

**Comment by StrahilKazlachev** [comment]
The latest is 1.2.1

**Comment by AshleyGrant** [comment]
https://github.com/aurelia/dependency-injection/commit/98739f2 and
https://github.com/aurelia/dependency-injection/commit/aa38a34
are the only changes that happened between 1.2.0 and 1.2.1

@StrahilKazlachev based on my rewatch of Rob explaining the `@singleton` decorator during the workshop he recorded in August, the behavior seen in `1.2.0` is the expected behavior.

**Comment by EisenbergEffect** [comment]
So, 1.2.1 isn't working then? That's going to be an interesting problem if it's true. Because that was a fix for singleton in another scenario.

**Comment by EisenbergEffect** [comment]
@AshleyGrant Can you link your test project here? I want to pull it down and take a look.

**Comment by StrahilKazlachev** [comment]
@EisenbergEffect [repro](https://github.com/AshleyGrant/decorator_bug_demo)

**Comment by EisenbergEffect** [comment]
Ok, I've got the fix for this. It was a bit tricky. I'll merge it and release tomorrow hopefully.

**Comment by EisenbergEffect** [comment]
Fixed and released.

**Comment by sorenhoyer** [comment]
There's still a problem with the singleton decorator in the latest aurelia-esnext-webpack-skeleton. I furthermore tried updating to the latest version 1.1.0 of "aurelia-framework", and checked in my nodes folder and npm dependency list that the "aurelia-dependency-injection" was indeed version 1.3.0. Still I get this error: http://aurelia-test.sorenhoyer.com/ - using the @singleton decorator it should not alert "parent" in the 3rd alert. Here's how it should work: https://gist.run/?id=c529ed1092d399dbab19d9eebdab3277

**Comment by sorenhoyer** [comment]
@AshleyGrant could I maybe get you to re-open the issue?

@EisenbergEffect would you perhaps take a look at this again? It seems to only be a problem for the child view which is initially loaded. I've written the steps necessary to replicate here: http://aurelia-test.sorenhoyer.com/ - hopefully that'll help narrow down the problem.

**Comment by AshleyGrant** [comment]
I'm unable to re-open it. @EisenbergEffect would you, please?

**Comment by sorenhoyer** [comment]
I just noticed (when trying in my large ui with lots of other routes outside the parent/child relationship in the example) that it is not linked to the **initial page load** - if that 1st rendered view is a totally different view which has nothing to do with the @singleton parent.

Instead this bug happens for the first "child" view/view-model that is activated. Navigating to the other children afterwards just fine.

**Comment by sorenhoyer** [comment]
I tried ALL released (systemjs/jspm esnext) versions since the ["layout-view"](https://github.com/aurelia/templating-router/commit/c5a0ad400a7abdcde16465cde831097d23645274) was introduced by @charlespockert in v. skeleton-navigation-1.0.0-beta.1.3.1 - without luck.

Then I tried to pull down @jdanyow 's [rjs-bundle](https://github.com/jdanyow/rjs-bundle), followed the instructions and updated and copyed over the  exact files from the [gist](https://gist.run/?id=c529ed1092d399dbab19d9eebdab3277). I only had to replace the script references in the index file to:

```
<script src="node_modules/requirejs/require.js"></script>
<script src="config.js"></script>
<script src="bundles/aurelia.js"></script>
<script src="bundles/babel.js"></script>
```

And then, after a "serve-export", everything worked just fine - just like in the gist. This was with the latest versions. 

So... the question is now:
Why does this work when using jeremy's rjs-bundle? And why doesn't it work when using esnext (jspm or webpack) skeletons? the aurelia.js is minified with jspm/webpack and with jeremy's bundle it isn't.

I had to manually copy them into the export folder

**Comment by Gheoan** [comment]
Perhaps because `rjs-bundle` is using old versions of aurelia libraries while the skeleton has the latest versions?

**Comment by AshleyGrant** [comment]
Yeah, that's probably why. There's been a regression at some point.

**Comment by jdanyow** [comment]
The rjs gist uses an old version of babel- not sure if that's a factor

**Comment by sorenhoyer** [comment]
Nope that was the point I tried to make. When I used rjs-bundle I did the step "Execute the following command to update all the aurelia libs:" so I had the latest versions. The aurelia versions doesn't seem to be the problem (also all versions since the introduction of "layout-view" had the problem, which just confirms it). The problem seem to be with the bundling in the jspm/webpack skeletons.

**Comment by sorenhoyer** [comment]
Could this problem be explained with https://github.com/loganfsmyth/babel-plugin-transform-decorators-legacy#best-effort ?

Would it be possible to achieve the same as https://gist.run/?id=c529ed1092d399dbab19d9eebdab3277 - but without the use of the @singleton decorator?

If we would somehow come up with the equivalent code **without the singleton decorator**, we would know for sure whether the problem is indeed the changed behavior regarding decorators in babel 6 (and if if works, my presentaiton wont fail miserably on monday heh ;-p)

Looking in the aurelia docs I took a deep look at Container, Resolver and how registerSingleton works, but so far I haven't been able to make it work.

**Comment by jdanyow** [comment]
@sorenhoyer I answered stackoverflow post- is that post the same as this issue?

http://stackoverflow.com/questions/42856797/refresh-bindings-in-parent-view-when-parent-value-is-changed-from-child-through/43018901#43018901

**Comment by sorenhoyer** [comment]
Yes, my focus on this issue came from that.

Your answer is great, explains the details and does give a good alternative and practical solution.

But wouldn't you still agree it's still an issue that we see different behavior when applying the @singleton decorator in babel 5 vs 6? As I see it, the gist (using your rjs-bundle) is the expected behavior and updating the aurelia dependencies of your bundle while seeing the same behavior just shows that this is not aurelia in itself. It can almost only be babel.

I wonder if this is something that requires a fix in the babel-plugin-transform-decorators-legacy or should the fix be made in Aurelia?

**Comment by jdanyow** [comment]
Possibly, although I think you'd need to use `@singleton(false)` instead of `@singleton(true)`.

Here's a diagram of what I think is happening, admittedly, I didn't step through everything again, but here's what I'm thinking:

Legend:
```
  [some text]: the name of a container instance
  {some text}: name of class instance that registered in container instance
```

### using `@singleton(true)` (registerInChild === true)

- [root container]
  - [child container used to construct ChildA]
    - {ChildA}
    - {SharedParent}
  - [child container used to construct the layout view model]
    - {SharedParent}

### using `@singleton(false)` (registerInChild === false)

- [root container]
  - {SharedParent}
  - [child container used to construct ChildA]
    - {ChildA}
  - [child container used to construct the layout view model]


You can verify/debug this by adding a `debugger` line to your class constructors and taking a look at the calls stack to determine why they're being constructed.

**Comment by sorenhoyer** [comment]
I just tested with the latest esnext-webpack skeleton. 

It seems to have been fixed. singleton(true) seem to reset the parent's bindings when navigating from 1 child to another, whereas singleton(false) seem to persist the bound values of the parent, when navigating between child routes.

At least for my use case, this is exactly how I would expect it to work.

The example can be found here: https://github.com/sorenhoyer/aurelia-singleton-parent-child-test-20170426 - it's exactly equivalent to the old one, posted above - only difference is the aurelia-skeleton version.
